js/plugins/google-cloud/src/telemetry/engagement.ts

```diff
@@ -0,0 +1,115 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Attributes, ValueType } from '@opentelemetry/api';
+import { ReadableSpan } from '@opentelemetry/sdk-trace-base';
+import { logger } from 'genkit/logging';
+import { PathMetadata } from 'genkit/tracing';
+import {
+  MetricCounter,
+  Telemetry,
+  internalMetricNamespaceWrap,
+} from '../metrics';
+import { createCommonLogAttributes } from '../utils';
+
+class EngagementTelemetry implements Telemetry {
+  /**
+   * Wraps the declared metrics in a Genkit-specific, internal namespace.
+   */
+  private _N = internalMetricNamespaceWrap.bind(null, 'engagement');
+
+  private feedbackCounter = new MetricCounter(this._N('feedback'), {
+    description: 'Counts calls to genkit flows.',
+    valueType: ValueType.INT,
+  });
+
+  private acceptanceCounter = new MetricCounter(this._N('acceptance'), {
+    description: 'Tracks unique flow paths per flow.',
+    valueType: ValueType.INT,
+  });
+
+  tick(
+    span: ReadableSpan,
+    paths: Set<PathMetadata>,
+    logIO: boolean,
+    projectId?: string
+  ): void {
+    const subtype = span.attributes['genkit:metadata:subtype'] as string;
+
+    if (subtype === 'userFeedback') {
+      this.writeUserFeedback(span, projectId);
+      return;
+    }
+
+    if (subtype === 'userAcceptance') {
+      this.writeUserAcceptance(span, projectId);
+      return;
+    }
+
+    logger.warn(`Unknown user engagement subtype: ${subtype}`);
+  }
+
+  private writeUserFeedback(span: ReadableSpan, projectId?: string) {
+    const attributes = span.attributes;
+    const name = this.extractTraceName(attributes);
+
+    const dimensions = {
+      name,
+      value: attributes['genkit:metadata:feedbackValue'],
+      hasText: !!attributes['genkit:metadata:textFeedback'],
+    };
+    this.feedbackCounter.add(1, dimensions);
+
+    const metadata = {
+      ...createCommonLogAttributes(span, projectId),
+      feedbackValue: attributes['genkit:metadata:feedbackValue'],
+    };
+    if (attributes['genkit:metadata:textFeedback']) {
+      metadata['textFeedback'] = attributes['genkit:metadata:textFeedback'];
+    }
+    logger.logStructured(`UserFeedback[${name}]`, metadata);
+  }
+
+  private writeUserAcceptance(span: ReadableSpan, projectId?: string) {
+    const attributes = span.attributes;
+    const name = this.extractTraceName(attributes);
+
+    const dimensions = {
+      name,
+      value: attributes['genkit:metadata:acceptanceValue'],
+    };
+    this.acceptanceCounter.add(1, dimensions);
+
+    const metadata = {
+      ...createCommonLogAttributes(span, projectId),
+      acceptanceValue: attributes['genkit:metadata:acceptanceValue'],
+    };
+    logger.logStructured(`UserAcceptance[${name}]`, metadata);
+  }
+
+  private extractTraceName(attributes: Attributes) {
```

Comment:
I don't think this can be directly re-used. `extractOuterFlowNameFromPath` explicitly looks for the `t:flow` portion of path, so this would only work if that's how we formatted our input (and we don't know if the feedback is for a flow). We can't change that function's regex to be more generic, because we don't want it to match anything that is not a flow (eg. it's also called from `generate.ts`).

I'm going to skip this refactor for now.